core/shared/src/main/scala/zio/stm/TSemaphore.scala

```diff
@@ -18,186 +18,53 @@ package zio.stm
- * larger STM transactions, for example acquiring permits from two different
- * semaphores transactionally and later releasing them transactionally to safely
- * synchronize on access to two different mutable variables.
- */
-final class TSemaphore private (val permits: TRef[Long]) extends Serializable {
-
-  /**
-   * Acquires a single permit in transactional context.
-   */
-  def acquire: USTM[Unit] =
-    acquireBetween(1L, 1L).unit
-
-  /**
-   * Acquires the specified number of permits in a transactional context.
-   */
-  def acquireN(n: Long): USTM[Unit] =
-    acquireBetween(n, n).unit
-
-  /**
-   * Acquire at least `min` permits and at most `max` permits in a transactional
-   * context.
-   */
-  def acquireBetween(min: Long, max: Long): USTM[Long] =
-    ZSTM.Effect { (journal, _, _) =>
-      require(min <= max, s"Unexpected `$min` > `$max` passed to acquireRange.")
-      assertNonNegative(min)
-      assertNonNegative(max)
-
-      val available: Long = permits.unsafeGet(journal)
-      if (available < min) {
-        throw ZSTM.RetryException
-      }
-      val requested: Long = available.min(max)
-      permits.unsafeSet(journal, available - requested)
-      requested
+import java.util.concurrent.ConcurrentLinkedQueue
+
+final class TSemaphore private (
+  val permits: TRef[Long],
+  fairness: Boolean,
+  private val waiters: ConcurrentLinkedQueue[Promise[Nothing, Unit]]
+) extends Serializable {
+
+  def acquire: USTM[Unit] = acquireN(1L)
+
+  def acquireN(n: Long): USTM[Unit] = ZSTM.Effect { (journal, _, _) =>
+    assertNonNegative(n)
+    val current = permits.unsafeGet(journal)
+    if (current >= n) {
+      permits.unsafeSet(journal, current - n)
+    } else {
+      throw ZSTM.RetryException
     }
+  }
 
-  /**
-   * Acquire at most `max` permits in a transactional context.
-   */
-  def acquireUpTo(max: Long): USTM[Long] =
-    acquireBetween(0, max)
-
-  /**
-   * Returns the number of available permits in a transactional context.
-   */
-  def available: USTM[Long] =
-    permits.get
-
-  /**
-   * Releases a single permit in a transactional context.
-   */
-  def release: USTM[Unit] =
-    releaseN(1L)
-
-  /**
-   * Releases the specified number of permits in a transactional context
-   */
-  def releaseN(n: Long): USTM[Unit] =
-    ZSTM.Effect { (journal, _, _) =>
-      assertNonNegative(n)
-
-      val current = permits.unsafeGet(journal)
-      permits.unsafeSet(journal, current + n)
+  def release: USTM[Unit] = releaseN(1L)
+
+  def releaseN(n: Long): USTM[Unit] = ZSTM.Effect { (journal, _, _) =>
+    assertNonNegative(n)
+    val current = permits.unsafeGet(journal)
+    permits.unsafeSet(journal, current + n)
+    if (!fairness) {
+      // Unfair: Wake waiters immediately without FIFO
+      var loop = true
+      while (loop && permits.unsafeGet(journal) > 0 && !waiters.isEmpty) {
+        val waiter = waiters.poll()
+        if (waiter != null) waiter.succeedUnit.unsafe.run
```

Comment:
What happens when we wake something?

Reply:
Note that whatever is waiting on this promise can't rely on the permits actually being available until the current transaction has been committed (which may fail).